### PR TITLE
[TS migration] Migrate `withOnyx.js` to TypeScript

### PR DESCRIPTION
**Pull request by fabioh8010**
### Details

This PR migrates `withOnyx.js` to TypeScript.

### Related Issues

https://github.com/Expensify/App/issues/34343

### Automated Tests

N/A

### Manual Tests

Some common tasks on E/App:

1. Send a message to someone.
2. React to that message.
3. Reply to that message, thus creating a thread.
4. Submit or track an expense.
5. Open Settings screen and assert is loading.

### Author Checklist

- [x] I linked the correct issue in the `### Related Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/react-native-onyx/assets/20051562/6bbc8539-2312-4710-a209-425811dd4ae9

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/react-native-onyx/assets/20051562/eb05f54c-78ed-4167-9545-3efeb1bbea8a

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/react-native-onyx/assets/20051562/c24fb4fa-944a-41fe-84c4-59c0822098b1

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/react-native-onyx/assets/20051562/3ee5bd16-d024-4aa9-a40b-bb0a2b2a8eb3

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/react-native-onyx/assets/20051562/739ee245-b8cb-445a-99c0-05dfc2e389a6

https://github.com/Expensify/react-native-onyx/assets/20051562/d3c349da-12a4-45d4-a3e8-8cc8f79f4806

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/react-native-onyx/assets/20051562/757db1a3-0dc5-4f29-ba4e-edcfacd7c108

</details>

### CODE REVIEW

File: lib/withOnyx.tsx
Comment (blazejkustra) [inline lib/withOnyx.tsx:1]: Shouldn't we remove d.ts file?

File: lib/withOnyx.tsx
Comment (blazejkustra) [inline lib/withOnyx.tsx:9]: Is underscore needed or are you still working on removing it?

File: lib/withOnyx.tsx
Comment (blazejkustra) [inline lib/withOnyx.tsx:182]: 👀

**Review by blazejkustra — CHANGES_REQUESTED** [review]

**Comment by blazejkustra** [comment]
@fabioh8010 Hey! Is this ready for review? Should I have a look?

File: lib/withOnyx.tsx
Comment (fabioh8010) [reply lib/withOnyx.tsx:1]: I will remove it soon!

File: types/modules/react.d.ts
Comment (blazejkustra) [inline types/modules/react.d.ts:3]: Won't this affect repos that use Onyx? Is this wanted?

File: lib/utils.ts
Comment (blazejkustra) [inline lib/utils.ts:141]: Let's add some JSDoc here, or a comment with the source

File: lib/withOnyx.tsx
Comment (blazejkustra) [inline lib/withOnyx.tsx:1]: How about we change the structure a bit to reduce the size of this file? Instead of `withOnyx.tsx` we could create a catalog `withOnyx` with `types.ts`, `index.tsx`, and some `utils.ts` (if needed).

File: lib/withOnyx.tsx
Comment (blazejkustra) [inline lib/withOnyx.tsx:124]: Some types are missing JSDoc, do you think it makes sense to add comments here and others?

File: lib/withOnyx.tsx
Comment (blazejkustra) [inline lib/withOnyx.tsx:189]: withOnyx.d.ts defines a different return type, I think this is the typecheck problem that you are experiencing:
```ts
declare function withOnyx<TComponentProps, TOnyxProps>(
    mapping: {
        [TOnyxProp in keyof TOnyxProps]: OnyxPropMapping<TComponentProps, TOnyxProps, TOnyxProp> | OnyxPropCollectionMapping<TComponentProps, TOnyxProps, TOnyxProp>;
    },
    shouldDelayUpdates?: boolean,
): (component: React.ComponentType<TComponentProps>) => React.ComponentType<Omit<TComponentProps, keyof TOnyxProps>>;
```

File: lib/withOnyx.tsx
Comment (blazejkustra) [inline lib/withOnyx.tsx:196]:
```suggestion
    const requiredKeysForInit = Object.keys(
        utils.omit(mapOnyxToState, ([, options]) => (options as MapOnyxToState<TComponentProps, TOnyxProps>[keyof TOnyxProps]).initWithStoredValues === false),
    );
```

File: lib/withOnyx.tsx
Comment (blazejkustra) [inline lib/withOnyx.tsx:460]:
```suggestion
                const onyxMapping = mapOnyxToState[statePropertyName] as WithOnyxMapping<TComponentProps, TOnyxProps>;
```

File: lib/withOnyx.tsx
Comment (blazejkustra) [inline lib/withOnyx.tsx:472]: I think we should adjust `WithOnyxInstance` type. `withOnyx` === `WithOnyxInstance`:
```ts
class withOnyx extends React.Component<WithOnyxProps<TComponentProps, TOnyxProps>, WithOnyxState<TOnyxProps>> {

type WithOnyxInstance = Component<unknown, WithOnyxInstanceState<NullableKeyValueMapping>> & {
    setStateProxy: (cb: (state: Record<string, OnyxCollection<KeyValueMapping[OnyxKey]>>) => OnyxValue<OnyxKey>) => void;
    setWithOnyxState: (statePropertyName: OnyxKey, value: OnyxValue<OnyxKey>) => void;
};
```

It may not be a trivial thing to do, so if you think it's not worth the effort you can resolve the comment.

**Review by blazejkustra — CHANGES_REQUESTED** [review]
I'm continuing review...

File: lib/withOnyx.tsx
Comment (blazejkustra) [inline lib/withOnyx.tsx:350]: Let's remove these lines 😄

File: lib/withOnyx.tsx
Comment (blazejkustra) [inline lib/withOnyx.tsx:509]:
```suggestion
                        {...(propsToPass as TComponentProps)}
```

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: lib/withOnyx.tsx
Comment (fabioh8010) [reply lib/withOnyx.tsx]: Good catch! Thanks!!

File: types/modules/react.d.ts
Comment (fabioh8010) [reply types/modules/react.d.ts]: I tested here and it won't affect, btw the .d.ts files won't be on `dist` folder anymore so it won't be copied.

File: lib/withOnyx.tsx
Comment (fabioh8010) [reply lib/withOnyx.tsx:1]: Good idea, I just didn't create `utils.ts` (seemed unnecessary to me)

File: lib/withOnyx.tsx
Comment (fabioh8010) [reply lib/withOnyx.tsx]: Yeah I tried for a couple hours to get the typings working, but I had a lot of issues on `OnyxUtils`, not sure if it's worth it.. but I did some improvements

File: lib/utils.ts
Comment (JKobrynski) [inline lib/utils.ts:168]: I think that you could use `Object.fromEntries` for this, but I'm not insisting 😄 

```ts
const result = Object.fromEntries(
    entries.filter(([key, value]) => {
        let shouldInclude;

        if (Array.isArray(condition)) {
            shouldInclude = condition.includes(key);
        } else if (typeof condition === 'string') {
            shouldInclude = key === condition;
        } else {
            shouldInclude = condition([key, value]);
        }

        return include ? shouldInclude : !shouldInclude;
    })
);
```

**Review by JKobrynski — APPROVED** [review]
Only one comment from me, LGTM!

File: lib/withOnyx/index.tsx
Comment (VickyStash) [inline lib/withOnyx/index.tsx:279]: It looks like it should work as well
```suggestion
                    const canEvict = !!Str.result(mapping.canEvict as GenericFunction, this.props);
```

File: lib/utils.ts
Comment (fabioh8010) [reply lib/utils.ts:168]: Yeah I chose the classic for-loop for performance reasons

File: lib/withOnyx/index.tsx
Comment (VickyStash) [inline lib/withOnyx/index.tsx:317]: Does it work this way?
```suggestion
                    withOnyxInstance: this as WithOnyxInstance,
```

**Review by VickyStash — APPROVED** [review]

**Review by blazejkustra — APPROVED** [review]
LGTM! Let's get 100% in Onyx! 🚀

File: lib/withOnyx/index.tsx
Comment (fabioh8010) [reply lib/withOnyx/index.tsx:317]: Unfortunately no 🙁 [I spent some time on it yesterday but no luck so far](https://github.com/Expensify/react-native-onyx/pull/542#discussion_r1603824466)

**Review by NikkiWines — APPROVED** [review]
Looks good, approving but not merging for now due to [this](https://expensify.slack.com/archives/C01GTK53T8Q/p1715880123893389)

**Comment by NikkiWines** [comment]
Yeah, since this is a bit on the higher-risk side for introducing front-end bugs we're gonna hold off on merging until the merge freeze is fully lifted. But don't worry it hasn't been forgotten about 😄

**Pull request merged**
